### DISCUSSION TITLE
0032. 最长有效括号

### DISCUSSION THREAD

**Discussion by giscus** [itcharge/AlgoNote#194] [Comments]
# 0032. 最长有效括号

[https://algo.itcharge.cn/solutions/0001-0099/longest-valid-parentheses/](https://algo.itcharge.cn/solutions/0001-0099/longest-valid-parentheses/)

**Comment by YangYishe** [comment]
关于栈思路的思考，如何理解这里的stack[-1]?
由于之前已经进行了pop操作，此处的stack[-1]记录的不是)对应的(的位置，而是这之前。可能为(的位置-1。
如果将()视为一层层嵌套的关系，则stack[-1]记录的则是当前层级()的最早计算位置
也因此，需要一个哨兵标记为第一层级的开始计算位置。使用栈进行计算的时候，涉及层级的概念时，均可考虑增加这么一个哨兵，来表示层级的开始计算位置
在此例中，尽管有层级的概念，但最后不一定能构建完整的层级。但这不影响利用层级栈
